src/controllers/import.js

```diff
@@ -106,25 +105,118 @@ function ImportController(context) {
+            // Concatenate each chunk of data into the buffer.
+            // Future performance enhancement, should issues arise: keep the buffer in memory
+            // and stream the data directly to S3 once jobId is created.
+            fileBuffer = Buffer.concat([fileBuffer, data]);
+          }).on('close', () => {
+            // Reached the end of the file data
+            log.debug(`File upload completed for ${name}`);
+            importScript = fileBuffer.toString('utf8');
+          });
+        }
+      });
+
+      // Handle other fields
+      busboy.on('field', (name, val) => {
+        try {
+          if (name === 'urls') {
+            urls = JSON.parse(val);
+          } else if (name === 'options') {
+            options = JSON.parse(val);
+          } else if (name === 'customHeaders') {
+            customHeaders = JSON.parse(val);
+          }
+          // Otherwise: ignore the field
+        } catch (error) {
+          reject(new ErrorWithStatusCode(`Unable to parse request data field (${name}): ${error.message}`, STATUS_BAD_REQUEST));
+        }
+      });
+
+      // Handle the end of the request
+      busboy.on('close', () => {
+        const requestData = {
+          urls,
+          options,
+          importScript,
+          customHeaders,
+        };
+        resolve(requestData);
+      });
+
+      busboy.on('error', (error) => {
+        reject(new ErrorWithStatusCode(`Invalid request, unable to parse request body: ${error.message}`, STATUS_BAD_REQUEST));
+      });
+
+      httpRequest.pipe(busboy);
+    });
+  }
+
   /**
    * Create and start a new import job.
-   * @param {object} requestContext - Context of the request.
-   * @param {Array<string>} requestContext.data.urls - Array of URLs to import.
-   * @param {object} requestContext.data.options - Optional import configuration parameters.
-   * @param {string} requestContext.pathInfo.headers.x-api-key - API key to use for the job.
+   * This function requires the Request object in order to parse the multipart/form-data mime-type
+   * of the request (required for large file uploads). The request is passed as a parameter to the
+   * controller, so we do not require any additional parameters to invoke this function.
    * @returns {Promise<Response>} 202 Accepted if successful, 4xx or 5xx otherwise.
    */
   async function createImportJob(requestContext) {
-    const { data, pathInfo: { headers } } = requestContext;
+    const { pathInfo: { headers } } = requestContext;
     const { 'x-api-key': importApiKey, 'user-agent': userAgent } = headers;
+
     try {
       // The API scope imports.write is required to create a new import job
       validateImportApiKey(importApiKey, ['imports.write']);
-      validateRequestData(data);
 
-      let initiatedBy = {};
+      // Parse the multipart request, which can include a custom import.js script as a file upload
+      const data = await parseMultipartRequest(request, headers);
```

Comment:
For backwards compatibility we should be handling the existing way that @arumsey and Labs is sending requests.  Anyone with postbuster scripts etc would be broken.  Essentially we have broken our API contract.  You would know better than me and the number of customers we have (which isn't that many I'm assuming), so I guess at this point we can communicated with the changes.

Reply:
Good thing we don't have many clients at this point that we can get away with this.  I agree that getting this updated before the announcement tomorrow and more uptake is a good idea. I just want to call out that changes like these going forward have to be looked with a fine tooth comb by the team.